P0 build blocker: lfortran/build-liric runtime generation aborts with free(): invalid pointer

## Summary
`../lfortran/build-liric` build fails in runtime generation because the just-built `lfortran` binary aborts while compiling runtime Fortran modules.

## Repro (2026-02-10)

```bash
cmake --build ../lfortran/build-liric -j$(nproc)
```

Failure stage:

```text
[12/16] Building Fortran object ... lfortran_intrinsic_iso_fortran_env.f90.o
/home/ert/code/lfortran-dev/lfortran/build-liric/src/bin/lfortran ... -c .../lfortran_intrinsic_iso_fortran_env.f90
free(): invalid pointer
ninja: build stopped: subcommand failed.
```

## Impact
- Cannot produce a healthy WITH_LIRIC lfortran toolchain.
- Blocks API benchmarking and runtime-dependent validation.

## Likely dependency
Depends on fixing startup/ABI crash path (see startup SIGABRT issue).

## Acceptance criteria
- `cmake --build ../lfortran/build-liric -j$(nproc)` completes runtime generation successfully.
- runtime `.mod`/object generation no longer aborts with `free(): invalid pointer`.

--- a/include/llvm/ADT/StringRef.h
+++ b/include/llvm/ADT/StringRef.h
@@ -19,20 +19,26 @@ class LIRIC_LLVM_HIDDEN StringRef {
 
 public:
     StringRef() : data_(""), len_(0) {}
-    StringRef(const char *s) : data_(s), len_(s ? std::strlen(s) : 0) {}
-    StringRef(const char *s, size_t n) : data_(s), len_(n) {}
+    StringRef(const char *s) : data_(s ? s : ""), len_(s ? std::strlen(s) : 0) {}
+    StringRef(const char *s, size_t n) : data_(s ? s : ""), len_(s ? n : 0) {}
     StringRef(const std::string &s) : data_(s.data()), len_(s.size()) {}
-    StringRef(std::string_view sv) : data_(sv.data()), len_(sv.size()) {}
+    StringRef(std::string_view sv)
+        : data_(sv.data() ? sv.data() : ""), len_(sv.data() ? sv.size() : 0) {}
 
     const char *data() const { return data_; }
     size_t size() const { return len_; }
     bool empty() const { return len_ == 0; }
 
-    std::string str() const { return std::string(data_, len_); }
+    std::string str() const {
+        if (len_ == 0) return std::string();
+        return std::string(data_, len_);
+    }
     operator std::string() const { return str(); }
 
     bool equals(StringRef other) const {
-        return len_ == other.len_ && std::memcmp(data_, other.data_, len_) == 0;
+        if (len_ != other.len_) return false;
+        if (len_ == 0) return true;
+        return std::memcmp(data_, other.data_, len_) == 0;
     }
     bool operator==(StringRef other) const { return equals(other); }
     bool operator!=(StringRef other) const { return !equals(other); }

--- a/include/llvm/ADT/Twine.h
+++ b/include/llvm/ADT/Twine.h
@@ -94,7 +94,7 @@ class LIRIC_LLVM_HIDDEN Twine {
             return;
         case PtrAndLengthKind:
         case StringLiteralKind:
-            if (node.ptrAndLength.ptr != nullptr || node.ptrAndLength.length == 0) {
+            if (node.ptrAndLength.ptr != nullptr && node.ptrAndLength.length != 0) {
                 out.append(node.ptrAndLength.ptr, node.ptrAndLength.length);
             }
             return;

--- a/include/llvm/Support/raw_ostream.h
+++ b/include/llvm/Support/raw_ostream.h
@@ -29,9 +29,11 @@ public:
         return write(&c, 1);
     }
     raw_ostream &operator<<(const std::string &s) {
+        if (s.empty()) return *this;
         return write(s.data(), s.size());
     }
     raw_ostream &operator<<(StringRef s) {
+        if (s.empty()) return *this;
         return write(s.data(), s.size());
     }
     raw_ostream &operator<<(int v) {
@@ -118,6 +120,7 @@ public:
     }
 
     raw_ostream &write(const char *ptr, size_t size) override {
+        if (size == 0) return *this;
         fwrite(ptr, 1, size, f_);
         return *this;
     }
@@ -132,6 +135,7 @@ class LIRIC_LLVM_HIDDEN raw_string_ostream : public raw_pwrite_stream {
 public:
     explicit raw_string_ostream(std::string &s) : str_(s) {}
     raw_ostream &write(const char *ptr, size_t size) override {
+        if (size == 0) return *this;
         str_.append(ptr, size);
         return *this;
     }
@@ -151,6 +155,7 @@ public:
     explicit raw_svector_ostream(VecT &) {}
 
     raw_ostream &write(const char *ptr, size_t size) override {
+        if (size == 0) return *this;
         if (vec_)
             vec_->insert(vec_->end(), ptr, ptr + size);
         else

--- a/tests/test_llvm_compat.cpp
+++ b/tests/test_llvm_compat.cpp
@@ -732,6 +732,23 @@ static int test_stringref_twine() {
     return 0;
 }
 
+static int test_stringref_nullptr_zero_len_safety() {
+    llvm::StringRef empty_from_null(nullptr, 0);
+    TEST_ASSERT(empty_from_null.empty(), "nullptr+0 StringRef is empty");
+
+    std::string rendered = empty_from_null.str();
+    TEST_ASSERT(rendered.empty(), "nullptr+0 StringRef renders empty");
+
+    llvm::Twine tw(empty_from_null);
+    TEST_ASSERT(tw.str().empty(), "twine from nullptr+0 StringRef renders empty");
+
+    std::string out;
+    llvm::raw_string_ostream os(out);
+    os << empty_from_null;
+    TEST_ASSERT(out.empty(), "raw_string_ostream ignores nullptr+0 StringRef");
+    return 0;
+}
+
 static int test_twine_abi_layout_and_concat() {
     size_t expected_size = sizeof(void *) == 8 ? 40u : 20u;
     TEST_ASSERT_EQ(sizeof(llvm::Twine), expected_size, "twine ABI size");
@@ -1115,6 +1132,7 @@ int main() {
     fprintf(stderr, "Infrastructure tests:\n");
     RUN_TEST(test_llvm_version);
     RUN_TEST(test_stringref_twine);
+    RUN_TEST(test_stringref_nullptr_zero_len_safety);
     RUN_TEST(test_twine_abi_layout_and_concat);
     RUN_TEST(test_arrayref);
     RUN_TEST(test_apint_apfloat);

